src/printer/index.js

```diff
@@ -154,8 +155,153 @@ function printArguments(print, path, argsKey, defaultsKey) {
+function countNumPrecedingChars(text, start, char, skipChar) {
+  skipChar = skipChar || "";
+  let count = 0;
+  for (
+    let indexToCheckNext = start - 1;
+    indexToCheckNext >= 0;
+    indexToCheckNext--
+  ) {
+    if (text[indexToCheckNext] === char) {
+      count++;
+    } else if (text[indexToCheckNext] !== skipChar) {
+      // If this is neither the target char nor the char we should skip,
+      // then we're done.
+      break;
+    }
+  }
+  return count;
+}
+
+function isBorderingFunctionOrClass(prevNode, node) {
+  const nodeTypesWthiTwoBlankLinesOnBothSides = [
+    "ClassDef",
+    "FunctionDef",
+    "AsyncFunctionDef"
+  ];
+  return (
+    nodeTypesWthiTwoBlankLinesOnBothSides.includes(node.ast_type) ||
+    nodeTypesWthiTwoBlankLinesOnBothSides.includes(prevNode.ast_type)
+  );
+}
+
+function isImport(node) {
+  return node.ast_type === "Import" || node.ast_type === "ImportFrom";
+}
+
+function upperBound(current, bound) {
+  return current > bound ? bound : current;
+}
+
+function lowerBound(current, bound) {
+  return current < bound ? bound : current;
+}
+
+function adjustTopLevelNewlines(numPrecedingNewlines, prevNode, node) {
+  // Start by bounding the number of blank lines to 2
+  numPrecedingNewlines = upperBound(numPrecedingNewlines, 3);
+
+  // If this isn't the very first node, which two nodes the blank lines
+  // are between may affect how many blank lines we need.
+  if (prevNode !== null) {
+    if (isBorderingFunctionOrClass(prevNode, node)) {
+      // We need two blank lines.
+      numPrecedingNewlines = 3;
+    } else if (isImport(prevNode)) {
+      if (isImport(node)) {
+        // If we're between imports, bound to one blank line
+        numPrecedingNewlines = upperBound(numPrecedingNewlines, 2);
+      } else {
+        // Otherwise if this is after imports we should have one blank
+        // line.
+        numPrecedingNewlines = 2;
+      }
+    }
+  }
+  return numPrecedingNewlines;
+}
+
+function adjustIndentedNewlines(numPrecedingNewlines, prevNode, node) {
+  // Start by bounding the number of blank lines to 1
+  if (numPrecedingNewlines > 2) {
+    numPrecedingNewlines = 2;
+  }
+  if (prevNode !== null) {
+    if (isBorderingFunctionOrClass(prevNode, node)) {
+      // We need one blank line.
+      numPrecedingNewlines = 2;
+    }
+  }
+  return numPrecedingNewlines;
+}
+
+function printBody(path, print, isTopLevel) {
+  const text = path.stack[0].source;
+  const childDocs = path.map(print, "body");
+  const childNodes = path.map(path => path.getValue(), "body");
+  const parts = [];
+
+  for (let i = 0; i < childDocs.length; i++) {
+    const doc = childDocs[i];
+    const node = childNodes[i];
+    const prevNode = i > 0 ? childNodes[i - 1] : null;
+
+    let start = node.start;
+    // This is a hack to work around an issue in asttokens, until the fix is
+    // merged: https://github.com/gristlabs/asttokens/pull/14
```

Comment:
I think we can remove this hack and update the vendorised version of asttoken, what do you think?

Reply:
Oh yeah, I meant to do that! Thanks for pointing it out :) I just sent out a separate PR to update the library (#41).